src/Microsoft.AspNetCore.SignalR.Common/Internal/Formatters/BinaryMessageParser.cs

```diff
@@ -10,36 +10,47 @@ public static class BinaryMessageParser
     {
         public static bool TryParseMessage(ref ReadOnlyBuffer<byte> buffer, out ReadOnlyBuffer<byte> payload)
         {
-            long length = 0;
-            payload = default(ReadOnlyBuffer<byte>);
+            payload = default;
 
-            if (buffer.Length < sizeof(long))
+            if (buffer.IsEmpty)
             {
                 return false;
             }
 
-            // Read the length
-            length = buffer.Span.Slice(0, sizeof(long)).ReadBigEndian<long>();
+            var length = 0U;
+            var numBytes = 0;
 
-            if (length > Int32.MaxValue)
+            byte byteRead;
+            do
             {
-                throw new FormatException("Messages over 2GB in size are not supported");
+                byteRead = buffer.Span.Slice(numBytes, sizeof(byte)).Read<byte>();
```

Comment:
I'm worried about the efficiency of this.

Reply:
We could try reading up Min(len, 5) bytes instead of slicing one byte at a time...